tests/backend/zero_alloc_checker/test_inference.mli

```diff
@@ -0,0 +1,25 @@
+val[@zero_alloc] f_zero_alloc : 'a -> 'a
+val[@zero_alloc] f_alloc : 'a -> 'a * 'a
+
+module M_alloc_var : sig
+  val[@zero_alloc] f_alloc2 : 'a -> 'a * 'a
+end
+val f_call_var : 'a -> 'a * 'a
+
+val[@zero_alloc (arity 1)] f_arity_one : int -> int -> int
+
+val[@zero_alloc] f_shadow : int -> int * int
+val[@zero_alloc] f_shadow_alloc : int -> int * int
+val[@zero_alloc] f_shadow_alloc_both : int -> int * int
+
+val[@zero_alloc] f_basic_success : int -> unit
+val[@zero_alloc] f_basic_fail : 'a -> 'a * 'a
+
+val[@zero_alloc strict] f_strict_success : int -> int
+val[@zero_alloc strict] f_strict_fail : int -> unit
+
+val[@zero_alloc opt] f_opt_success : int -> unit
+val[@zero_alloc opt] f_opt_fail : int -> int * int
+
+val[@zero_alloc strict opt] f_strict_opt_success : int -> int
+val[@zero_alloc strict opt] f_strict_opt_fail : int -> unit
```

Comment:
I think we should add a test demonstrating the "spooky action at a distance" of a module inclusion check causing the zero-alloc check to flag an issue with the original module definition. (You have type-checking tests for something like this.)

e.g.
```ocaml
module type S = functor () -> sig val[@zero_alloc] f : 'a -> 'a option end

module F () = struct
  let f x = Some x
end

module _ : S = F
```

In particular, the error message does not identify why the zero-alloc check is running at all, and I think it would be good to have a test demonstrating the iffy error message (and even better to someday make it better).